### ISSUE TITLE
Ensuring objdictgen is in sync with the canfestival library variant used

### ISSUE THREAD

**Issue by sveinse** [Laerdal/python-objdictgen#8]
`objdictgen` is a generic tool for reading and writing `.od` files. It is also used to do code generation of the parameters. This code generation is very linked to the specific version or variant of the canfestival library used. Currently this tool is designed against https://github.com/Laerdal/canfestival-laerdal .

This tool is uploaded to pypi, which indicates that this is intended to be a generic tool that is available to others as well. AFAIK canfestival doesn't have a normative upstream repo, so it should be expected that there are many variants around for this repo.

How do we best make sure that the code generation by objdictgen is consistent with the canfestival library?

**Comment by robybona** [comment]
I have done a short analysis of the original canfestival repo (https://github.com/Laerdal/canfestival-3-asc.git) to try to understand where the code generation changed.

Looking at the `objdictgen/gen_cfile.py` history, it's possible to see that some commits are responsible for the generated .c and .h files broken compatibility.

The relevant ones are:
- 2014-05-01 `61663187e5144e0a02564554198595a859eaf7a3` Arrays of strings or domains are now declared as two dimensions arrays instead of arrays of pointers.
- 2014-05-09 `fc9c633e817b5bbbe29ef79e962020a627e3f4e1` Change of the scanIndexOD_t prototype : add of a CO_Data* d parameter.
- 2014-05-09 `537fe3bf579d2093f948659400dfa9228b35303f` Another change in scanIndexOD_t prototype, callbacks functions are now part of the subindex struct.

I think that the "legacy" file generation can be related to the source code up to the last 2013 commit:
- 2013-06-05 `84db3b03c3419e9b0cfffa577244ce80fb4ba718` merge with mainline dev.automforge.net/CanFestival-3/

The "standard" file generation can be related instead to the source code that modified the .c and .h output in a substantial way. So, from this commit (included):
- 2014-05-09 `537fe3bf579d2093f948659400dfa9228b35303f` Another change in scanIndexOD_t prototype, callbacks functions are now part of the subindex struct.

There are other changes to `objdictgen/gen_cfile.py` in 2015 and later on but, as long as I can understand, they didn't break the compatibility anymore.

In relation to the custom Laerdal repo (https://github.com/Laerdal/canfestival-laerdal.git), there are further changes to `objdictgen/gen_cfile.py` (after 2015) that should be evaluated to understand whether the compatibility has been kept or not.

So, to recap the situation, there should be two variants (or three, basing on the custom Laerdal repo evaluation result) for the file generation that could be managed to guarantee the compatibility all along the whole canfestival history.

The mentioned patch (https://github.com/Laerdal/python-objdictgen/pull/6) aims to cope with the first two variants (what has been called "legacy" and "standard"). Some tests on that patch showed that it is not fully functional, so it couldn't be merged as it is. Anyway, @aduskett could adjust the patch for the scope.

**Comment by sveinse** [comment]
@robybona cool. Thanks.

One thought that comes to mind is if it possible to appoint one canfestval repo as being the "official" canfestival repo. Without that, I think it will be hard to maintain "this is the correct way", as shown in this analysis.

A possible solution for dealing with varying canfestival repos, and thus `gen_cfile.py`, could be that the `objdictgen` tool is get support for code generation plugins. That `gen_cfile.py` follows the canfestival repo and that objdictgen can be configured to use the provided variant.

For this to work, the interface / API that the `gan_cfile.py` use must be standardized and versioned, but I think it could be doable.

**Comment by aduskett** [comment]
Having a plugin interface sounds like a fantastic idea. However, we would need to discuss how the plugin interface would be set up.

What I am imagining for an initial commit is something like:
  - Remove the existing code in the gen_cfile.py file.
  - Make a new “GenCFile” API class in the gen_cfile.py file.
  - Add a new laerdal.py file in a c_file_plugins directory that uses the GenCFile API class. 

Of course, this also means the following needs to be added:
  - A new plugin option in the GUI to load plugins (or a settings file to specify a plugins directory)
  - Proper method headers to create documentation for others to use the API
  - New unit and integration tests to ensure the API works properly.

Other questions:
  - What is the desired structure of the API?
  - Does the API use consistent formatting with Black, Flake8, and MyPi? Or does it conform to the way things look in the current code base?

Thanks!
Adam

**Comment by sveinse** [comment]
Without having looked at it yet, I think making `gen_cfile.py` as a plugin should be fairly simple. The only external function `objdictgen` use of this file is `GenerateFile()`, and I think it does so in fairly ancient versions of it as well. What might be missing here is versioning on how `gen_cfile.py` consumes the `Node` object. It have to be looked at.

With regards to formatting and and code base: `objdictgen` was until quite recently dual py2 and py3 code. That prevented us from using any of the modern py3 methods for coding, annotating and linting. Now that py2 support have been removed, its on my list to do a overhaul into more "pythonic" py3 code. Contributions are welcome!

**Comment by robybona** [comment]
@sveinse In my personal opinion, this issue should be refused and closed, at the end.
In fact, what I called "legacy" file generation here: https://github.com/Laerdal/python-objdictgen/issues/8#issuecomment-1902588064, is related to code that's more than 10 years old.
Moreover, I think it's far easier for developers using so old canfestival code to patch it to conform to the "standard" file generation, than burdens this project with managing the "legacy" one, in some way.
